streetcomplete/StreetComplete
# Surface overlay

![screen04](https://user-images.githubusercontent.com/899988/202921305-7c13d181-fe68-4769-be4a-27f314ecde0d.png)
fixes #4378

open questions:
Do we need "sidewalks are mapped here as separate lines" for places where sidewalks are mapped both as road property and separate lines? See https://www.openstreetmap.org/way/218801173 

Maybe list of surface should use entire screen and fit three in width.

Note: `highway=cycleway sidewalk=right` type of ways are NOT supported ( that is popular in some areas, see http://overpass-turbo.eu/s/1oem )

I tested following things which should work (I made this notes while handling some special cases, and before I submitted PR I went over and retested it again - fixing what was broken).
Let me know if you see anything where automatic testing can be reasonably added and I missed it.

- test are passing
- android tests are failing - but in the same way as on master (ActiveDatesDaoTest, worth looking into it but I want to slay this monster first - it is funny how much effort went into it, large part into all this :note corner case handling)
- save/restore of menu state
- editing road surface
- editing path surface

- adding value where nothing was previously tagged - or where surface is tagged already
- adding paved/unpaved
    - with long explanation
    - try to answer with empty notes
- regular value
- selecting the same stuff as tagged already

- editing path surface segregated into parts for cyclists and pedestrians
   - and handling surface tag
       - when both are the same
       - when both are paved
       - when both are unpaved
       - when one was paved and one is unpaved (no surface tagged at all)
       - above: also when surface is tagged already
   - where path was claiming to have both bicycles and pedestrians but was not marked as segregated=yes (switching via can't say)
   - editing one of already tagged values
- switching path surface into one segregated by cyclists and pedestrians
- handling surface of path (sidewalks) tagged as road property,in one layer with separately marked paths
   - roads that have no sidewalk present as tags cannot be edited
- handling unspecific surfaces with surface:note (not shown as missing data but with special style)
   - for paths
   - for roads
   - for paths, also when no surface tag is present (not eligible for overlay)
   - for paths, also when no surface tag is present (not eligible for overlay)
   - also for sidewalks tagged as property on the main road (6 such objects worldwide :) )
       - no support for editing them
- handling aliased surfaces (earth/dirt, cement/concrete) by showing them as aliased one
- handling unrecognised surfaces by blocking selection for overlay
- handling unspecific surfaces like surface=paved/unpaved/cobblestone etc (by treating it as a missing data and displaying it to user as missing data)
     - opening unspecific surface for editing and trying to exit menu should not be trigger confirmation about missing data
- cycleway:surface:note, footway:surface:note, surface:note with matching surface
   - shown to user, both on path and on road
        - editable, removed on editing, added when asked for on selecting surface=paved surface=unpaved
- cycleway:surface:note, footway:surface:note, surface:note without matching surface
        - surface:note without surface makes impossible to save changes to just note, you need select surface what resets note text (the same for cycleway: footway: prefixed) - only few ( https://overpass-turbo.eu/s/1nNg )
        - 0 cases for `footway:surface:note` without `footway:surface` https://overpass-turbo.eu/s/1rvA
        - likewise for cycleway https://overpass-turbo.eu/s/1rvB
   - show surface:note also where it is not paved/unpaved if set already
   - show surface:note without surface but with footway:surface
    - emptying surface:note should delete it (it is fine to allow it with surface=paved - it will just trigger a new quest)
- remove associated surface tags on changing surface (check_date:surface, colour:surface etc) - changes here affected also regular surface quest
- skip ways with foobar:surface=* / turbo_surface_tag=* and other keys with surface in its text - unless listed as handled (this may include removal on edit, see above)
- surface:note is editable
- cycleway:surface:note and footway:surface:note are also editable
- surface=soil surface=earth and other synonyms are well handled
- surface=cement will now be resurveyed by surface quest
- split enabled and working on `segregated=no` ways with bicycle access
   - splitting requires providing surface before saving
   - split not enabled without bicycle access
- sidewalk as tag on private roads without surface data does not show up as red http://overpass-turbo.eu/s/1pxb
- sidewalk as tag on private roads with surface data is shown (`highway=residential and access=private and foot!=* and sidewalk:both:surface=* in Warwick`)
- (not tested, should work) surface:note with footway:surface:note (does not happen worldwide http://overpass-turbo.eu/s/1nRj )
- (not tested, should work) footway:surface:note where it is not paved/unpaved - is it shown? (does not happen woldwide...)
- (not tested, should work) cycleway:surface:note where it is not paved/unpaved - is it shown? (does not happen woldwide...)
- (not tested) way with unlisted surface value for sidewalk is listed (with generic descriptio)
- way with unlisted surface value for path is shown (with generic description)
- way with unlisted surface value for road is shown (with generic description)
- things like `surface=paved;unpaved` treated as missing surface (as it is result of a badly done merges)
- invalid surface like `surface=cobblestone` / `surface=cement` works well when `surface:note` is set
- (not tested) surface:note without surface [effectively the same as above] - note `!surface:note or` query filters
- `sidewalk:both:surface` `sidewalk:left:surface` `sidewalk:right:surface` with weird values https://taginfo.openstreetmap.org/keys/sidewalk%3Aboth%3Asurface#values

Done now: <s>Wait for few days on  https://community.openstreetmap.org/t/mysterious-surface-cement/5158 and then edit surface=cement page documenting it as deprecated (and remove TODO in the code mentioning that surface=cement is bing treated as deprecated without wiki clearly describing it as deprecated)</s>

Rejected for now:

Do we need to add distinction between asphalt, concrete, paving_stones?
If yes: how we can achieve it? Suble changes of currently assigned colours? But between blue overlay colours differences are already barely noticeable (or maybe I am a bit colour blind for shades of blue? Is it a thing? Color.SKY and Color.BLUE seem extremely close to me)

I tried using dashing but I cannot get any different style to work. 
In `if (style.stroke.dashed) props["dashed"] = "1"` you can replace 1 with [1, 10] or "banan" and you will get the same effect. 
Appears to be effect of https://github.com/tangrams/tangram-es/issues/2023 and any advanced dash styling is not available.
Yeah, I should work more on MapLibre this 6 months ago rather than waste that opportunity.
Well.

Maybe surface=paved/unpaved should be shown to user? Rather than presented as empty data? (then what about surface=cobblestone?)

Done, thanks:

I would appreciate testing/review/feedback/nitpicking/complaints (pinging especially @FloEdelmann - maybe you have time and interest to look through)
